### DISCUSSION TITLE
Custom Entity

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#68] [Q&A]
Can I create custom types inheriting from `Db.Entity` that will write additional data to the dwg database. I see the `.dwgOutFields()` method in `ARX`, but I don't see it in `PyRx`.

**Comment by CEXT-Dan** [comment]
Unfortunately, no.  The goal is to be able to simulate one with overrules, but I only have ObjectOverrule and DrawableOverrule, though this is usually enough for most.

I’ve thought about some different options, such as making a couple of base custom objects that could be extended or overruled. I also need to consider if someone wants to add in a C++/Python custom object.  As of now, the only way to do this is fork the project

Is this for something specific? Maybe I can recommend an approach.  BTW, you can examine the contents of dwgOutFields with the snoop filer, see testDbObjectSnoop sample

**Reply by gswifort** [reply]
I want to store data used by my python modules in a dwg file. To simplify, let's assume that I am drawing a ceiling with a polyline and I want to save its concrete class. I haven't gone into detail yet, but I think I could use Xdata, Xrecord or Dictionary.

**Reply by CEXT-Dan** [reply]
Yes, you can serialize a class (pickle?)
For Xrecord, you can use kDxfBinaryChunk, and xdata kDxfXdBinaryChunk

It would be the same concept at this.
https://www.theswamp.org/index.php?topic=26687

The downside is, you must break data into 127byte chunks. 
If I get time, I will try to make a sample

**Reply by CEXT-Dan** [reply]
you will have to wait for the next release as I just implemented kDxfBinaryChunk and kDxfXdBinaryChunk
but here's a proof of concept 

```Python
import traceback
from pyrx_imp import Rx, Ge, Gi, Db, Ap, Ed
import pickle

class PyData:
    sd1 = ""
    sd2 = ""
    fd1 = 0.0
    id1 = 1

    def printData(self):
        print(self.sd1)
        print(self.sd2)
        print(self.fd1)
        print(self.id1)

def divide_chunks(l): 
    for i in range(0, len(l), 126):  
        yield l[i:i + 126] 
        
def xdbuilder(data):
    bOut = pickle.dumps(data)
    dx = list(divide_chunks(bOut)) 
    xd = [(Db.DxfCode.kDxfRegAppName, "PYTHONTEST")]
    for x in dx:
        xd.append((Db.DxfCode.kDxfXdBinaryChunk, x))
    return xd
        
def xdreader(xd):
    b = bytearray()
    for x in xd:
        if x[0] !=  Db.DxfCode.kDxfXdBinaryChunk:
            continue
        b.extend(x[1])
    return pickle.loads(b)

def PyRxCmd_doit() -> None:
    try:
        Db.Core.regApp("PYTHONTEST")

        esres = Ed.Editor.entSel("\nSelect: \n")
        if esres[0] != Ed.PromptStatus.eOk:
            raise RuntimeError("Entsel Error {}: ".format(esres[0]))

        data = PyData()
        data.sd1 = "check out the brain on brad, lets make this bigger than 127 bytes"
        data.sd2 = "check out the brain on brad, lets make this bigger than 127 bytes"
        data.fd1 = 3.14159265359
        data.id1 = 42
        
        xd = xdbuilder(data)
        ent = Db.Entity(esres[1], Db.OpenMode.kForWrite)
        ent.setXData(xd)
    except Exception as err:
        traceback.print_exception(err)


def PyRxCmd_doit2() -> None:
    try:
        esres = Ed.Editor.entSel("\nSelect: \n")
        if esres[0] != Ed.PromptStatus.eOk:
            raise RuntimeError("Entsel Error {}: ".format(esres[0]))
        ent = Db.Entity(esres[1])
        xd = ent.xData("PYTHONTEST")
        data = xdreader(xd)
        data.printData()

    except Exception as err:
        traceback.print_exception(err)
```

![bxdata](https://github.com/CEXT-Dan/PyRx/assets/26830697/18ceca6b-8acb-492d-998b-1ed4de605548)

**Reply by gswifort** [reply]
Looks like what I need. Thank you!
Can I easily filter all elements containing specific XData, e.g. "PYTHONTEST"?
Can this be done differently than by iterating the entire dwg database?

**Reply by CEXT-Dan** [reply]
one option would be to use a selection set filter
```Python
def PyRxCmd_doit3() -> None:
    try:
        filter = [(Db.DxfCode.kDxfXDataStart, 0),
            (Db.DxfCode.kDxfRegAppName, "PYTHONTEST")]
        ss = Ed.Editor.selectAll(filter)
        print(len(ss[1].objectIds()))
    except Exception as err:
        traceback.print_exception(err)
```

**Reply by gswifort** [reply]
I will be grateful for the above example also with `Xrecord`. As I understand, then I will not have data size limitation like in `Xdata`.

**Reply by CEXT-Dan** [reply]
I probably didn’t explain this well enough. the 127-byte limit is the maximum chunk size 
This is both Xrecord and Xdata, Xdata itself will hold like 16kb 

i've revised my sample

**Reply by CEXT-Dan** [reply]
Oh I found something even cooler! this creates an Xrecord

```Python
import traceback
from pyrx_imp import Rx, Ge, Gi, Db, Ap, Ed
import pickle

class PyData:
    sd1 = ""
    sd2 = ""
    fd1 = 0.0
    id1 = 1

    def printData(self):
        print("\n{}\n{}\n{}\n{}".format(self.sd1,self.sd2,self.fd1,self.id1))

def PyRxCmd_doit() -> None:
    try:
        esres = Ed.Editor.entSel("\nSelect: ")
        if esres[0] != Ed.PromptStatus.eOk:
            raise RuntimeError("Entsel Error {}: ".format(esres[0]))

        data = PyData()
        data.sd1 = "check out the brain on brad, lets make this bigger than 127 bytes"
        data.sd2 = "check out the brain on brad, lets make this bigger than 127 bytes"
        data.fd1 = 3.14159265359
        data.id1 = 42
        
        ent = Db.Entity(esres[1], Db.OpenMode.kForWrite)
        ent.setBinaryData("PYTHONTEST", pickle.dumps(data))
        
    except Exception as err:
        traceback.print_exception(err)

def PyRxCmd_doit2() -> None:
    try:
        esres = Ed.Editor.entSel("\nSelect: ")
        if esres[0] != Ed.PromptStatus.eOk:
            raise RuntimeError("Entsel Error {}: ".format(esres[0]))
        
        ent = Db.Entity(esres[1])
        bOut = ent.getBinaryData("PYTHONTEST")
        data = pickle.loads(bOut)
        
        data.printData()
        
    except Exception as err:
```
![xrec](https://github.com/CEXT-Dan/PyRx/assets/26830697/0b48deed-58b9-4776-a77e-43bdafba183d)

**Reply by CEXT-Dan** [reply]
I added functions setXDBinaryData and  getXDBinaryData that will behave exactly like  setBinaryData and getBinaryData shown above, but it's xdata.. I'll post a sample

**Comment by CEXT-Dan** [comment]
getBinaryData, setBinaryData ~~fail on ZwCAD~~, pass on AutoCAD, BricsCAD and GstarCAD
getXDBinaryData, setXDBinaryData works on all

**Reply by CEXT-Dan** [reply]
getBinaryData, setBinaryData getXDBinaryData, setXDBinaryData now pass all platforms

**Comment by gswifort** [comment]
As I understand it, `obj.setBinaryData()` adds an Xrecord object to the database and assigns it an `obj` id as a parent.  Could this method return the id of the Xrecord object so that nested structures can be created?

```py
xr1 = Db.Xrecord(ent.setBinaryData(...))
xr2 = Db.Xrecord(xr1.setBinaryData(...))
xr3 = Db.Xrecord(xr2.setBinaryData(...))
```

**Comment by CEXT-Dan** [comment]
Yes, I can try to make setBinaryData return the objectId of the Xrecord

However, if you notice the picture I posted, obj.setBinaryData creates a Dictionary with the obj as its parent, 
then adds an entry [KEY, Xrecord].  the owner of the Xrecord is the Dictionary. So, you can already store multiple Xrecords per entity.

![xr](https://github.com/CEXT-Dan/PyRx/assets/26830697/99595f03-1f3a-4ecf-8dd7-4b28f8d8e4d7)

I think with the nested approach, it may be difficult to query if the object has nested data, and creating a lot of dictionaries with one entry.
